benchmarks/benchmark_functions/qaoa.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Benchmarks for QAOA optimizations.
+"""
+import pennylane as qml
+from pennylane import qaoa
+from .default_settings import _qaoa_defaults
+
+
+def benchmark_qaoa(hyperparams={}):
+	"""
+	Performs QAOA optimizations.
+
+	Args:
+		hyperparams (dict): hyperparameters to configure this benchmark
+
+			* 'graph': Graph represented as a NetworkX Graph class
+
+			* 'n_layers': Number of layers in the QAOA circuit
+
+			* 'params': Numpy array of trainable parameters that is fed into the circuit
+
+			* 'n_steps': Number of QAOA steps
+
+			* 'device': Device on which the circuit is run, or valid device name
+
+			* 'interface': Name of the interface to use
+
+			* 'diff_method': Name of differentiation method
+	"""
+
+	graph, n_layers, params, n_steps, device, options_dict = _qaoa_defaults(hyperparams)
+
+	H_cost, H_mixer = qaoa.min_vertex_cover(graph, constrained=False)
+
+	wires = range(len(params[0]))
+
+	def qaoa_layer(gamma, alpha):
+		qaoa.cost_layer(gamma, H_cost)
+		qaoa.mixer_layer(alpha, H_mixer)
+
+	def circuit(params, **kwargs):
+		for w in wires:
+			qml.Hadamard(wires=w)
+		qml.layer(qaoa_layer, n_layers, params[0], params[1])
+
+	cost_fn = qml.ExpvalCost(circuit, H_cost, device, **options_dict)
+
+	opt = qml.GradientDescentOptimizer(stepsize=0.4)
+
+	for n in range(n_steps):
+		params = opt.step(cost_fn, params)
+
+	@qml.qnode(device)
+	def probability_circuit(gamma, alpha):
+		circuit([gamma, alpha])
+		return qml.probs(wires=wires)
+
+	probs = probability_circuit(params[0], params[1])
```

Comment:
This code was taken from the intro to QAOA tutorial, where it made sense to compute all probabilities. However, a more representative output of the algorithm is the value of the cost function. I would recommend changing the return to 
`qml.expval(H_cost)` and changing the name of the function to just `circuit`.

Alternatively, this could be a good opportunity to test running devices with a *finite number of shots*. In the algorithm, what you're actually doing is sampling outputs and keeping the best one, so it makes sense to generate many independent outputs

Reply:
Actually, it seems like `expval()` doesn't support passing Hamiltonian objects...

Maybe we could get the exp value of the operators that appear in `H_cost` manually?

Reply:
Not sure if this is coincidence, but @josh146  told me yesterday that this would be really nice to support, and probably not too difficult...just as a side note.

Reply:
Yep

Reply:
> This code was taken from the intro to QAOA tutorial, where it made sense to compute all probabilities. However, a more representative output of the algorithm is the value of the cost function. I would recommend changing the return to
> `qml.expval(H_cost)` and changing the name of the function to just `circuit`.
> 
> Alternatively, this could be a good opportunity to test running devices with a _finite number of shots_. In the algorithm, what you're actually doing is sampling outputs and keeping the best one, so it makes sense to generate many independent outputs

How about `expval = cost_fn(params)`?

Reply:
So what I'd like to suggest is to follow the approach of the [QAOA for MaxCut](https://pennylane.ai/qml/demos/tutorial_qaoa_maxcut.html) demo. 

Here we use a device with a finite number of shots, e.g.,
```python
dev = qml.device("default.qubit", wires=n_wires, shots=1)
```

and define an identity operator as the observable we are sampling from
```python
def comp_basis_measurement(wires):
    n_wires = len(wires)
    return qml.Hermitian(np.diag(range(2 ** n_wires)), wires=wires)
```
and then return
```python
return qml.sample(comp_basis_measurement(range(n_wires)))
```
in the qnode. We can then benchmark how long it takes to generate one QAOA sample

Reply:
This approach is actually something I would have liked to 'fix' in PennyLane, since `qml.Hermitian` is _very_ inefficient.

It would be better to instead do:

```python
return [qml.sample(qml.PauliZ(i)) for i in range(n_wires)]
```
to extract a list of samples, and then do post-processing if you need the decimal representation of the computational basis state:

```python
import pennylane as qml
import numpy as np

dev = qml.device('default.qubit', wires=3)

@qml.qnode(dev)
def circuit(x, y, z):
	qml.RX(x, wires=0)
	qml.RX(x, wires=1)
	qml.RX(x, wires=2)
	qml.CNOT(wires=[0, 1])
	qml.CNOT(wires=[1, 2])
	qml.CNOT(wires=[2, 1])
	return [qml.sample(qml.PauliZ(i)) for i in range(3)]

samples = circuit(0.3, 0.1, 0.4).T
samples = (1 - samples) / 2  # convert from (1, -1) to (0, 1)
print(samples)

# convert to decimal representation
two_powers = 2 ** np.arange(samples.shape[-1])[::-1]
decimals = samples @ two_powers
print(decimals)
```

Reply:
As a better, more long-term solution, I would like to add `return qml.sample_basis()` to PL to automate the process of returning computational basis samples.